core/emulate.c

```diff
@@ -0,0 +1,807 @@
+    op->mem.seg = SEG_DS;
+    if (ctxt->override_segment) {
+        op->mem.seg = ctxt->override_segment;
+    }
+}
+
+/* Soft-emulation */
+static void em_mov(struct em_context_t *ctxt)
+{
+    memcpy(&ctxt->dst, &ctxt->src1, ctxt->operand_size);
+}
+
+static void em_movzx(struct em_context_t *ctxt)
+{
+    uint64_t value = 0;
+    memcpy(&value, &ctxt->src1, ctxt->operand_size);
+    ctxt->dst.value = value;
+}
+
+static void em_xchg(struct em_context_t *ctxt)
+{
+    uint64_t src1, src2;
+    src1 = ctxt->src1.value;
+    src2 = ctxt->src2.value;
+    ctxt->src1.value = src2;
+    ctxt->src2.value = src1;
+    operand_write(ctxt, &ctxt->src1);
+    operand_write(ctxt, &ctxt->src2);
+}
+
+em_status_t em_decode_insn(struct em_context_t *ctxt, uint8_t *insn)
+{
+    uint8_t b;
+    uint64_t flags;
+    struct em_opcode_t *opcode;
+    const struct em_opcode_t *opcode_group;
+
+    switch (ctxt->mode) {
+    case EM_MODE_PROT16:
+        ctxt->operand_size = 2;
+        ctxt->address_size = 2;
+        break;
+    case EM_MODE_PROT32:
+        ctxt->operand_size = 4;
+        ctxt->address_size = 4;
+        break;
+    case EM_MODE_PROT64:
+        ctxt->operand_size = 4;
+        ctxt->address_size = 8;
+        break;
+    default:
+        return EM_ERROR;
+    }
+    ctxt->override_segment = SEG_NONE;
+    ctxt->override_operand_size = 0;
+    ctxt->override_address_size = 0;
+    ctxt->insn = insn;
+    ctxt->lock = 0;
+    ctxt->rep = 0;
+    ctxt->len = 0;   
+    decode_prefixes(ctxt);
+
+    /* Apply legacy prefixes */
+    if (ctxt->override_operand_size) {
+        ctxt->operand_size ^= (2 | 4);
+    }
+    if (ctxt->override_address_size) {
+        ctxt->address_size ^= (ctxt->mode != EM_MODE_PROT64) ? (2 | 4) : (4 | 8);
+    }
+
+    /* Intel SDM Vol. 2A: 2.2.1 REX Prefixes */
+    ctxt->rex.value = 0;
+    b = insn_fetch_u8(ctxt);
+    if (ctxt->mode == EM_MODE_PROT64 && b >= 0x40 && b <= 0x4F) {
+        ctxt->rex.value = b;
+        if (ctxt->rex.w) {
+            ctxt->operand_size = 8;
+        }
+        b = insn_fetch_u8(ctxt);
+    }
+
+    /* Intel SDM Vol. 2A: 2.1.2 Opcodes */
+    opcode = &ctxt->opcode;
+    *opcode = opcode_table[b];
+    if (b == 0x0F) {
+        b = insn_fetch_u8(ctxt);
+        switch (b) {
+        case 0x38:
+            b = insn_fetch_u8(ctxt);
+            *opcode = opcode_table_0F38[b];
+        case 0x3A:
+            b = insn_fetch_u8(ctxt);
+            *opcode = opcode_table_0F3A[b];
+
+        default:
+            *opcode = opcode_table_0F[b];
+        }
+    }
+
+    /* Intel SDM Vol. 2A: 2.1.3 ModR/M and SIB Bytes */
+    flags = opcode->flags;
+    if (flags & INSN_MODRM) {
+        ctxt->modrm.value = insn_fetch_u8(ctxt);
+    }
+
+    /* Apply flags */
+    if (flags & INSN_BYTEOP) {
+        ctxt->operand_size = 1;
+    }
+    if (flags & INSN_GROUP) {
+        opcode_group = &opcode->group[ctxt->modrm.opc];
+        opcode->handler = opcode_group->handler;
```

Comment:
> Should make sure `opcode_group != NULL`.

Well `opcode_group = &opcode->group[ctxt->modrm.opc];` is always non-NULL because `opcode->group` is always non-NULL (otherwise it would make no sense to use `INSN_GROUP`).

However, you are completely right that opcode_group could point to a zero-initialized entry with a `NULL` handler (e.g. in case of unimplemented/unexpected ModRM opcodes), so I added a check for `opcode_group->handler != NULL`, and if it fails, the emulator decoder will throw `EM_ERROR`.

> IIUC, for unsupported ModR/M opcodes, you just initialize the corresponding group members to `NULL`

The opcode group is initialized as:
```c
static const struct em_opcode_t opcode_groupN[8] = {
    ...
};
```

AFAIK, C99 ensures that any element not explicitly defined (e.g. via `F(em_*, op_*, ...)`) will be zero-initialized.

Reply:
You're right. I missed the `&` operator earlier ;-)

Also guess there's not much benefit of initializing opcode group entries to the `N` macro, because `handler` would still be `NULL`. BTW, currently `N` is almost the same as all-zero, since the `INSN_NOTIMPL` flag is not checked anywhere.

nit: Move declaration of `opcode_group` into this `if` block.